### PR TITLE
Enable formula input and output validation.

### PR DESCRIPTION
**Pull request by Laimiux**
## What

I've added a verification/validation mode that enables us to ensure that equality guarantees are fulfilled by formula implementations. This mode will always run `Formula.evaluate` twice and check that the output/input/actions do not change for identical runs.

### CODE REVIEW

File: formula/src/test/java/com/instacart/formula/FormulaRuntimeTest.kt [hunk 1]
```diff
@@ -175,7 +175,7 @@ class FormulaRuntimeTest(val runtime: TestableRuntime, val name: String) {
             .apply { formula.incrementEvents.triggerEvent() }
             .apply { formula.incrementEvents.triggerEvent() }
             .apply {
-                val expected = listOf(0, 0, 1, 2, 3)
+                val expected = listOf(0, 1, 2, 3)
```
Comment (Laimiux) [inline formula/src/test/java/com/instacart/formula/FormulaRuntimeTest.kt:178]: Made the output a data class, so now it doesn't emit duplicate outputs.

File: formula/src/test/java/com/instacart/formula/subjects/DuplicateListenerKeysHandledByIndexing.kt [hunk 2]
```diff
@@ -10,7 +10,7 @@ object DuplicateListenerKeysHandledByIndexing {
 
     fun test(runtime: TestableRuntime) = runtime.test(TestFormula(), Unit)
 
-    class TestOutput(
+    data class TestOutput(
```
Comment (Laimiux) [inline formula/src/test/java/com/instacart/formula/subjects/DuplicateListenerKeysHandledByIndexing.kt:13]: A bunch of places were captured by the validation logic which didn't provide equality implementation of the output.

**Comment by carrotkite** [comment]
### JaCoCo Code Coverage 79.05% :white_check_mark:
| Class | Covered | Meta | Status |
|:---|:---:|:---:|:---:|
| `com/instacart/formula/coroutines/FlowRuntime` | 100% | 0% | :white_check_mark: |
| `com/instacart/formula/internal/FormulaManagerImpl` | 79% | 0% | :white_check_mark: |
| `com/instacart/formula/internal/ValidationException` | 0% | 0% | :white_check_mark: |
| `com/instacart/formula/rxjava3/RxJavaRuntime` | 100% | 0% | :white_check_mark: |
| `com/instacart/formula/FormulaRuntime` | 82% | 0% | :white_check_mark: |
| `com/instacart/formula/test/RxJavaFormulaTestDelegate` | 100% | 0% | :white_check_mark: |


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by rjaylwar — APPROVED** [review]
LGTM

**Pull request merged**